test/test_selectors.py

```diff
@@ -0,0 +1,670 @@
+        self.assertGreaterEqual(get_time() - t, LONG_SELECT - TOLERANCE)
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_interrupt_with_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+        key = s.get_key(rd)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, lambda *args: wr.send(b'x'))
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        t = get_time()
+        self.assertEqual([(key, selectors.EVENT_READ)], s.select(LONG_SELECT))
+        self.assertLess(get_time() - t, LONG_SELECT)
+        self.assertEqual(rd.recv(1), b'x')
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_multiple_interrupts_with_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+        key = s.get_key(rd)
+
+        def second_alarm(*args):
+            wr.send(b'x')
+
+        def first_alarm(*args):
+            self.set_alarm(SHORT_SELECT)
+            signal.signal(signal.SIGALRM, second_alarm)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, first_alarm)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        t = get_time()
+        self.assertEqual([(key, selectors.EVENT_READ)], s.select(LONG_SELECT))
+        t = get_time() - t
+        self.assertGreaterEqual(t, SHORT_SELECT * 2)
+        self.assertLess(t, LONG_SELECT)
+        self.assertEqual(rd.recv(1), b'x')
+
+    # Test ensures that _syscall_wrapper properly raises the
+    # exception that is raised from an interrupt handler.
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_interrupt_exception(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+
+        class AlarmInterrupt(Exception):
+            pass
+
+        def alarm_exception(*args):
+            raise AlarmInterrupt()
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, alarm_exception)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        t = get_time()
+        self.assertRaises(AlarmInterrupt, s.select, LONG_SELECT)
+        self.assertLess(get_time() - t, LONG_SELECT)
+
+    def test_fileno(self):
+        s = self.make_selector()
+        if hasattr(s, "fileno"):
+            fd = s.fileno()
+            self.assertTrue(isinstance(fd, int))
+            self.assertGreaterEqual(fd, 0)
```

Comment:
This test probably *should* formally skip itself if the selector does not have a `fileno` attribute.